src/components/rule-builder/rule-builder.component.tsx

```diff
@@ -0,0 +1,1746 @@
+} from '../../constants';
+import styles from './rule-builder.scss';
+import dayjs from 'dayjs';
+import { findQuestionIndices } from '../utils';
+
+export interface Condition {
+  id: string;
+  isNew: boolean;
+  logicalOperator?: string;
+  targetField?: string;
+  targetCondition?: string;
+  targetValue?: string;
+  targetValues?: Array<{ concept: string; label: string }>;
+}
+
+export interface Action {
+  id: string;
+  isNew: boolean;
+  logicalOperator?: string;
+  actionCondition?: string;
+  actionField?: string;
+  calculateField?: string;
+  errorMessage?: string;
+}
+
+export interface FormRule {
+  id: string;
+  question: string;
+  conditions?: Array<Condition>;
+  actions?: Array<Action>;
+  isNewRule?: boolean;
+  validatorIndex?: number;
+}
+
+interface RuleBuilderProps {
+  ruleId: string;
+  key: string;
+  question: Question;
+  pageIndex: number;
+  sectionIndex: number;
+  questionIndex: number;
+  onSchemaChange: (schema: Schema) => void;
+  isNewRule: boolean;
+  schema: Schema;
+  handleAddLogic: (fieldId: string) => void;
+}
+
+const RuleBuilder = React.memo(
+  ({
+    ruleId,
+    schema,
+    onSchemaChange,
+    question,
+    pageIndex,
+    sectionIndex,
+    questionIndex,
+    isNewRule,
+    handleAddLogic,
+  }: RuleBuilderProps) => {
+    const pages: Array<Page> = schema?.pages || [];
+    const sections: Array<Section> = pages.flatMap((page) => page.sections || []);
+    const questions: Array<Question> = sections.flatMap((section) => section.questions || []);
+    const answers: Array<Record<string, string>> = questions.flatMap((question) => question.questionOptions.answers);
+    const isTablet = useLayoutType() === 'tablet';
+    const responsiveSize = isTablet ? 'lg' : 'sm';
+
+    const { rules, setRules } = useFormRule();
+
+    const [conditions, setConditions] = useState<Array<Condition>>([{ id: uuidv4(), isNew: false }]);
+    const [actions, setActions] = useState<Array<Action>>([{ id: uuidv4(), isNew: false }]);
+    const [currentRule, setCurrentRule] = useState<FormRule>({
+      id: uuidv4(),
+      question: question.id,
+      actions: actions,
+      conditions: conditions,
+      isNewRule: false,
+    });
+    const [isRequired, setIsRequired] = useState<boolean>(question?.required ? true : false);
+    const [isDisallowDecimals, setIsDisallowDecimals] = useState<boolean>(question?.questionOptions?.disallowDecimals);
+    const [isHistoryEnable, setIsHistoryEnable] = useState<boolean>(
+      question?.historicalExpression?.length > 0 || false,
+    );
+    const [isAllowFutureDate, setIsAllowFutureDate] = useState<boolean>(
+      question?.validators?.some((validator) =>
+        validator.type === RenderingType.DATE && validator?.allowFutureDates === 'true' ? true : false,
+      ),
+    );
+    const [ruleDeleted, setRuleDeleted] = useState<boolean>(false);
+
+    const prevPageIndex = useRef(-1);
+    const prevSectionIndex = useRef(-1);
+    const prevQuestionIndex = useRef(-1);
```

Comment:
I'm a little concerned that these are really intended to be _states_ rather than _refs_.